src/models/StyledComponent.js

```diff
@@ -16,138 +16,186 @@ export default (ComponentStyle: Function, constructWithOptions: Function) => {
+        const generatedClassName = this.generateAndInjectStyles(theme, nextProps)
+
+        return { theme, generatedClassName }
+      })
+    }
 
-      componentWillUnmount() {
-        if (this.unsubscribe) {
-          this.unsubscribe()
-        }
+    componentWillUnmount() {
+      if (this.unsubscribe) {
+        this.unsubscribe()
       }
+    }
+
+    render() {
+      const { children, innerRef } = this.props
+      const { generatedClassName } = this.state
+      const { styledComponentId, target } = this.constructor
+
+      const isTargetTag = isTag(target)
+
+      const className = [
+        this.props.className,
+        styledComponentId,
+        this.attrs.className,
+        generatedClassName,
+      ].filter(Boolean).join(' ')
+
+      const propsForElement = Object
+        .keys(this.props)
+        .reduce((acc, propName) => {
+          // Don't pass through non HTML tags through to HTML elements
+          if (
+            !isTargetTag || (
+              validAttr(propName) &&
+              propName !== 'className' &&
+              propName !== 'innerRef'
+            )
+          ) {
+            // eslint-disable-next-line no-param-reassign
+            acc[propName] = this.props[propName]
+          }
+
+          return acc
+        }, {
+          ...this.attrs,
+          className,
+          ref: innerRef,
+          innerRef: undefined,
+        })
+
+      return createElement(target, propsForElement, children)
+    }
+  }
 
-      render() {
-        const { className, children, innerRef } = this.props
-        const { generatedClassName } = this.state
-
-        const propsForElement = { ...this.attrs }
-        /* Don't pass through non HTML tags through to HTML elements */
-        Object.keys(this.props)
-          .filter(propName => !isTag(target) || validAttr(propName))
-          .forEach(propName => {
-            propsForElement[propName] = this.props[propName]
-          })
-        propsForElement.className = [className, componentId, this.attrs.className, generatedClassName].filter(x => x).join(' ')
-        if (innerRef) {
-          propsForElement.ref = innerRef
-          delete propsForElement.innerRef
-        }
-
-        if (typeof process !== 'undefined' && process.env.NODE_ENV !== 'production' && generatedClassName) {
-          warnTooManyClasses(generatedClassName, StyledComponent.displayName)
-        }
-        return createElement(target, propsForElement, children)
+  const createStyledComponent = (
+    target: Target,
+    options: Object,
+    rules: RuleSet,
+  ) => {
+    const {
+      displayName = isTag(target) ? `styled.${target}` : `Styled(${target.displayName})`,
+      componentId = generateId(options.displayName),
+      ParentComponent = BaseStyledComponent,
+      rules: extendingRules,
+      attrs,
+    } = options
+
+    const warnTooManyClasses = createWarnTooManyClasses(displayName)
```

Comment:
seems like this should be guarded for dev-mode only - outside the scope of this PR perhaps, but it does introduce a bit of overhead and library size